src/libs/common/types/TextLine.ts

```diff
@@ -0,0 +1,47 @@
+import type { Range } from "@cursorless/common";
+
+/**
+ * Represents a line of text, such as a line of source code.
+ *
+ * TextLine objects are __immutable__. When a {@link TextDocument document} changes,
+ * previously retrieved lines will not represent the latest state.
+ */
+export interface TextLine {
+  /**
+   * The zero-based line number.
+   */
+  readonly lineNumber: number;
+
+  /**
+   * The text of this line without the line separator characters.
+   */
+  readonly text: string;
+
+  /**
+   * The range this line covers without the line separator characters.
+   */
+  readonly range: Range;
+
+  /**
+   * The range this line covers with the line separator characters.
+   */
+  readonly rangeIncludingLineBreak: Range;
+
+  /**
+   * The offset of the first character which is not a whitespace character as defined
+   * by `/\s/`. **Note** that if a line is all whitespace the length of the line is returned.
+   */
+  readonly firstNonWhitespaceCharacterIndex: number;
+
+  /**
+   * The offset of the last character which is not a whitespace character as defined
+   * by `/\s/`. **Note** that if a line is all whitespace the length of the line is returned.
```

Comment:
Correct. We kinda had both of these flipped around, but because the way the range constructed works it turned out correctly. I have now made a commit to fix this.

Reply:
Tbh I think I'd go with the reverse: if it is all whitespace then firstNonWhitespace should be length, and lastNonWhitespace should be 0.  That seems more logical: eg if client forgets to check if all whitespace and tries to make a range of the line's "content", then your current impl will end up with them getting a range full of whitespace, whereas this proposal will result in empty range, which seems like what you'd want.  It also means we don't need special case logic in the implementation; that will just happen automatically

Reply:
That's not true.  Let's say I do something like:

```typescript
const contentRange = line.range.with(new Position(line.lineNumber, line.firstNonWhitespaceCharacterIndex));
const content = document.getText(contentRange)
```

In your current impl, you'll get all whitespace.  With my impl, you'd get an empty string.  Same goes if you do the corresponding thing with `lastNonWhitespaceCharacterIndex`:

```typescript
const contentRange = line.range.with(undefined, new Position(line.lineNumber, line.lastNonWhitespaceCharacterIndex));
const content = document.getText(contentRange)
```

Reply:
I think you're talking about

```typescript
new Range(line.lineNumber, line.firstNonWhitespaceCharacterIndex, line.lineNumber, line.lastNonWhitespaceCharacterIndex)
```

which is fair, but I don't think detracts from my example above.  Tbh it's tempting not to support auto-swapping start and end for that reason, but it is useful in other scenarios

I wonder if we should have a property `trimmedRange` that just would return empty range at start of line if line is all whitespace.

Reply:
Please go ahead